### DISCUSSION TITLE
Prerendering Blazor Apps - How does it work / tips and tricks

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#18] [General]
# Prerendering Blazor Apps - How does it work / tips and tricks

Blazor comes with the option to prerender your webpage on the server. This works for the client-side version as well as the server side version of Blazor.
Let's have a look how does it work and what might be some pitfalls and how can we come around those.

https://steven-giesel.com/blogPost/aa5a6a5c-7f73-4262-9acb-14b645f79417

**Comment by cdomino** [comment]
Nice post! However, there is a typo in the code above; you need to remove the `!` from the `if` statement, as `TryTakeFromJson` returns `true` if the "take" operation was successful.

Thanks! 
Domino

**Reply by linkdotnet** [reply]
Thanks for reporting !

**Reply by linkdotnet** [reply]
Corrected

**Reply by devdestech** [reply]
Hello I hope you're very well.
Congrats for your post, its amazing.

Friend, I have a component, and I call 2 endpoints. 
I followed the post for my first endpoint and this one persist, a.i..

protected override async Task OnInitializedAsync()
    {
        persistingSubscription = applicationState.RegisterOnPersisting(PersistData);
        if (applicationState.TryTakeFromJson<IEnumerable<ApplicationUserDto>>(cachingKey, out var restored))
            accountTenants = restored;
        else
            accountTenants = await accountSearchTenant.Execute("*");

....
But, 

I want to persist my other endpoint. How Can I achieve that?
For example
 if (applicationState.TryTakeFromJson<PaginationProductDto>(cachingKey, out var restoredos))
        {
            paginationProductDto = restoredos;
        }
        else
        {
            pagination = new PaginationDto();
            paginationProductDto = new PaginationProductDto();
            await this.GetData();
        }

That second case, when the component refresh, that one is executed in the "ELSE". its never execute 

Can you help me, please?

**Reply by linkdotnet** [reply]
Hmm do you have a complete repository for that to check that out? I am also not certain what you mean with 2nd enpoint in the context of Blazor. Do you mean 2nd page which tries to persist a different state?
It is quite hard to debug that from that code snippet, as there are multiple things going on.

For example: Is the caching key is "wrong" (or even the same key). The ApplicationState is singleton, so your two pages would share the same instance. Therefore they need a different key.

**Comment by TicuIonutbe** [comment]
Hi, I believe u will have the good answer for my question on stackoverflow.
Will u be kind to help me with an answer please?
https://stackoverflow.com/questions/74760239/blazor-state-is-fe-or-be

I would really really appreciate.
I just want to know if there are 2 requests from the user to the server.
My question is explained on stackoverflow.

**Reply by linkdotnet** [reply]
Hey @TicuIonutbe 

I just quote my own answer from SO:
> No, the saved application state is part of the HTML. Oversimplified the state you persist is JSON-encoded inside the HTML the server sends to the client.<br/><br/>
Check the HTML code if your application with a persisted application state and look for your properties ;)
